mathics_scanner/tokeniser.py

```diff
@@ -159,57 +173,85 @@
     ("Del", r" \u2207 "),
     # uf520 is Wolfram custom, 25ab is standard unicode
     ("Square", r" \uf520 | \u25ab"),
-    ("SmallCircle", r" \u2218 "),
-    ("CircleDot", r" \u2299 "),
     # ('Sum', r' \u2211 '),
     # ('Product', r' \u220f '),
-    ("PlusMinus", r" \u00b1 "),
-    ("MinusPlus", r" \u2213 "),
     ("Nor", r" \u22BD "),
     ("Nand", r" \u22BC "),
     ("Xor", r" \u22BB "),
     ("Xnor", r" \uF4A2 "),
-    ("Diamond", r" \u22c4 "),
-    ("Wedge", r" \u22c0 "),
-    ("Vee", r" \u22c1 "),
-    ("CircleTimes", r" \u2297 "),
-    ("CenterDot", r" \u00b7 "),
-    ("Star", r" \u22c6"),
-    ("VerticalTilde", r" \u2240 "),
-    ("Coproduct", r" \u2210 "),
-    ("Cap", r" \u2322 "),
-    ("Cup", r" \u2323 "),
-    ("CirclePlus", r" \u2295 "),
-    ("CircleMinus", r" \u2296 "),
-    ("Congruent", r" \u2261 "),
     ("Intersection", r" \u22c2 "),
     ("Union", r" \u22c3 "),
-    ("VerticalBar", r" \u2223 "),
-    ("NotVerticalBar", r" \u2224 "),
-    ("DoubleVerticalBar", r" \u2225 "),
-    ("NotDoubleVerticalBar", r" \u2226 "),
     ("Element", r" \u2208 "),
     ("NotElement", r" \u2209 "),
-    ("Subset", r" \u2282 "),
-    ("Superset", r" \u2283 "),
     ("ForAll", r" \u2200 "),
     ("Exists", r" \u2203 "),
     ("NotExists", r" \u2204 "),
     ("Not", r" \u00AC "),
     ("Equivalent", r" \u29E6 "),
     ("Implies", r" \uF523 "),
-    ("RightTee", r" \u22A2 "),
-    ("DoubleRightTee", r" \u22A8 "),
-    ("LeftTee", r" \u22A3 "),
-    ("DoubleLeftTee", r" \u2AE4 "),
-    ("SuchThat", r" \u220D "),
     ("VerticalSeparator", r" \uF432 "),
-    ("Therefore", r" \u2234 "),
-    ("Because", r" \u2235 "),
-    ("Backslash", r" \u2216 "),
 ]
 
 
+def compile_pattern(pattern):
+    return re.compile(pattern, re.VERBOSE)
+
+
+def compile_tokens(token_list):
+    return [(tag, compile_pattern(pattern)) for tag, pattern in token_list]
+
+
+def find_indices(token_list: list, literals: dict) -> dict:
+    """Return a list of indices for literal tokens taking into account
+    the contents of token_list"""
+
+    literal_indices = {}
+    for key, tags in literals.items():
+        indices = []
+        for tag in tags:
+            for i, (tag2, _) in enumerate(token_list):
+                if tag == tag2:
+                    indices.append(i)
+                    break
+        assert len(indices) == len(
+            tags
+        ), f"problem matching tokens for symbol {key} having tags {tags}"
+        literal_indices[key] = tuple(indices)
+    return literal_indices
+
+
+# Initalized below in update_tokens_from_JSON
+tokens = []
```

Comment:
Are these variables a part of the module interface? Or are they just internal variables?

Reply:
tokens and literal tokens start out as tuples of operator name and string value and are used to create position indexes which are used in parsing. 

but after this, they are converted to some sort of compiled expression, and then I think these type-changed variables are used in parsing.  (It may have been in some style of programming where variables morph after their old value is no longer needed, was possibly at one time considered clever. Nowadays, it is considered annoying and frowned upon. In fact. in strongly-typed languages you are not allowed to do this.)

I have split out those two uses. We could run `del tokens` when they are no longer needed, but, right now, I don't think it is worth the bother.